sources/api/apiclient/src/exec.rs

```diff
@@ -0,0 +1,683 @@
+    }
+}
+
+/// ReadFromUser is responsible for reading user input from stdin and sending it to the given
+/// channel so it can be forwarded to the server.
+struct ReadFromUser {}
+
+impl ReadFromUser {
+    /// Parameters:
+    /// * stdin_tx: The channel to which we should send messages containing user input.
+    ///
+    /// * capacity_reader: We'll only read input when the server has capacity, according to this
+    /// parameter, so that we don't unnecessarily fill buffers or overwhelm the server.
+    ///
+    /// * is_tty: whether input is coming from a TTY; think of it as whether the command is
+    /// interactive.  If so, we read a byte at a time and send it immediately to the server so that
+    /// things like tab completion work.
+    fn new(
+        stdin_tx: mpsc::UnboundedSender<Message>,
+        capacity_reader: Arc<AtomicCapacity>,
+        is_tty: bool,
+    ) -> Self {
+        debug!("Spawning thread to read from user");
+        let stdin_fn = if is_tty {
+            Self::read_stdin_tty
+        } else {
+            Self::read_stdin
+        };
+        thread::spawn(move || {
+            if let Err(e) = stdin_fn(stdin_tx, capacity_reader) {
+                error!("Read from user failed: {}", e);
+            }
+        });
+
+        Self {}
+    }
+
+    /// Read from stdin with the lowest possible latency, sending each byte to the server.
+    fn read_stdin_tty(
+        tx: mpsc::UnboundedSender<Message>,
+        capacity: Arc<AtomicCapacity>,
+    ) -> Result<()> {
+        let mut stdin = std::io::stdin();
+        // Keep track of the number of messages we've read.  We compare this to the number of
+        // messages the server has written, as received in its regular capacity update messages, so
+        // that we don't overwhelm the server.
+        let mut messages_read = 0u64;
+
+        loop {
+            // Wait for server to have capacity for writes before reading; don't give "false hope" to
+            // whatever's writing to our stdin that it'll be read until there's room for it.
+            // (Note: we're unlikely to hit this interactively, which is the primary use for TTY.)
+            Self::wait_for_capacity(messages_read, &capacity);
+
+            // Read a byte at a time.
+            let mut buf = [0; 1];
+            match stdin.read_exact(&mut buf) {
+                Ok(()) => {
+                    messages_read += 1;
+                    // Send the data to the server in a Binary message without encoding.
+                    tx.unbounded_send(Message::Binary(Vec::from(buf)))
+                        .context(error::SendMessage { kind: "user input" })?;
+                }
+                // We don't normally get Err, since the server will close connection first, but for
+                // completeness...
+                Err(e) => {
+                    if e.kind() == std::io::ErrorKind::UnexpectedEof {
+                        debug!("Finished reading input");
```

Comment:
We don't normally get EOF because the server sends us a Close first, so we'll hit the `select!` and exit the process before this thread sees an EOF... and yeah, maybe if the user ends their TTY?  But I think other things would end it first in that case...